demo/api/src/open-telemetry/api-metrics.interceptor.ts

```diff
@@ -0,0 +1,59 @@
+import { CallHandler, ExecutionContext, Injectable, NestInterceptor } from "@nestjs/common";
+import { GqlExecutionContext } from "@nestjs/graphql";
+import { Counter, Histogram } from "@opentelemetry/api";
+import { GraphQLResolveInfo } from "graphql";
+import { Observable, tap } from "rxjs";
+
+import { getOrCreateCounter, getOrCreateHistogram } from "./metrics";
+
+@Injectable()
+export class ApiMetricsInterceptor implements NestInterceptor {
+    private gqlResolverCount: Counter;
+    private resolverDuration: Histogram;
+
+    constructor() {
+        this.gqlResolverCount = getOrCreateCounter("gql.resolver.count", {
+            description: "Total number of Root GraphQL resolver calls",
+            unit: "requests",
+        });
+        this.resolverDuration = getOrCreateHistogram("gql.resolver.duration", {
+            description: "The duration Root GraphQL resolver calls",
+            unit: "ms",
+        });
+    }
+
+    private isResolvingGraphQLField(gqlInfo: GraphQLResolveInfo): boolean {
+        const parentType = gqlInfo.parentType.name;
+        return parentType !== "Query" && parentType !== "Mutation";
+    }
+    public intercept(context: ExecutionContext, next: CallHandler): Observable<unknown> {
+        const requestType = context.getType().toString();
+        let gqlInfo: GraphQLResolveInfo;
+        if (requestType === "graphql") {
+            const graphqlExecutionContext = GqlExecutionContext.create(context);
+            gqlInfo = graphqlExecutionContext.getInfo<GraphQLResolveInfo>();
+
+            if (this.isResolvingGraphQLField(gqlInfo)) {
+                return next.handle();
+            }
+        } else {
+            return next.handle();
```

Comment:
Should we add metrics for controllers as well? Or is this handle by the ApiMetricsMiddleware?

Reply:
yes, it is handled by the middleware.